File: algorithms/0017-letter-combinations-of-a-phone-number.py
class Solution:
    def letterCombinations(self, digits: str) -> List[str]:
        chs = {
            '2': 'abc',
            '3': 'def',
            '4': 'ghi',
            '5': 'jkl',
            '6': 'mno',
            '7': 'pqsr',
            '8': 'tuv',
            '9': 'wxyz',
        }
        out = list()
        for i in digits:
            cur = list()
            if i == '1':
                return []
            for c in chs[i]:
                if not out:
                    cur.append(c)
                for e in out:
                    cur.append(e + c)
            out = cur
        return out

File: algorithms/0017-letter-combinations-of-a-phone-number.go
package algorithms

func letterCombinations(digits string) []string {
	var out []string
	for _, i := range digits {
		var cur, cs []string
		switch i {
		case '1':
			return cur
		case '2':
			cs = []string {"a", "b", "c"}
		case '3':
			cs = []string {"d", "e", "f"}
		case '4':
			cs = []string {"g", "h", "i"}
		case '5':
			cs = []string {"j", "k", "l"}
		case '6':
			cs = []string {"m", "n", "o"}
		case '7':
			cs = []string {"p", "q", "r", "s"}
		case '8':
			cs = []string {"t", "u", "v"}
		case '9':
			cs = []string {"w", "x", "y", "z"}
		}
		for _, c := range cs {
			if len(out) == 0 {
				cur = append(cur, c)
			}
			for _, e := range out {
				cur = append(cur, e + c)
			}
		}
		out = cur
	}
	return out
}
